lib/connextclient/ConnextClient.ts

```diff
@@ -289,15 +289,47 @@ class ConnextClient extends SwapClient {
         amount: unitsToRequest.toLocaleString('fullwide', { useGrouping: false }),
       }).then(() => {
         this.logger.debug(`completed collateral request of ${unitsToRequest} ${currency} units`);
-        this.requestCollateralPromises.delete(currency);
         return this.channelBalance(currency);
-      }).catch(this.logger.error);
+      }).catch(this.logger.error).finally(() => {
+        this.requestCollateralPromises.delete(currency);
+      });
       this.requestCollateralPromises.set(currency, requestCollateralPromise);
 
       throw errors.INSUFFICIENT_COLLATERAL;
     }
   }
 
+  public setReservedInboundAmount = (reservedInboundAmount: number, currency: string) => {
+    const inboundCapacity = this._maxChannelInboundAmount.get(currency) || 0;
+    if (inboundCapacity < reservedInboundAmount) {
+      // we do not have enough inbound capacity to fill all open orders, so we will request more
+      this.logger.debug(`collateral of ${inboundCapacity} for ${currency} is insufficient for reserved order amount of ${reservedInboundAmount}`);
+
+      // we want to make a request for the current collateral plus the greater of any
+      // minimum request size for the currency or the capacity shortage + 3% buffer
+      const quantityToRequest = inboundCapacity + Math.max(
+        reservedInboundAmount * 1.03 - inboundCapacity,
+        ConnextClient.MIN_COLLATERAL_REQUEST_SIZES[currency] ?? 0,
+      );
+      const unitsToRequest = this.unitConverter.amountToUnits({ currency, amount: quantityToRequest });
+
+      // first check whether we already have a pending collateral request for this currency
+      // if not start a new request, and when it completes call channelBalance to refresh our inbound capacity
+      // we don't await this request - instead we allow for "lazy collateralization" to complete since
+      // we don't expect all orders to be filled at once, we can be patient
+      const requestCollateralPromise = this.requestCollateralPromises.get(currency) ?? this.sendRequest('/request-collateral', 'POST', {
```

Comment:
This logic below seems to be the same as the one defined in `public checkInboundCapacity`. Can we extract this to a function?

Reply:
Done.